statsmodels/formula/_manager.py

```diff
@@ -0,0 +1,723 @@
+
+    Parameters
+    ----------
+    engine : {"patsy", "formulaic"} or None
+        The formula engine to use. If None, the default engine, which appears
+        in the attribute statsmodels.formula.formula_options.engine, is used.
+
+    Raises
+    ------
+    ValueError
+        If the selected engine is not available.
+    """
+
+    def __init__(self, engine: Literal["patsy", "formulaic"] | None = None):
+        self._engine = self._get_engine(engine)
+        self._spec = None
+        self._missing_mask = None
+
+    def _get_engine(
+        self, engine: Literal["patsy", "formulaic"] | None = None
+    ) -> Literal["patsy", "formulaic"]:
+        """
+        Check user engine selection or get the default engine to use.
+
+        Parameters
+        ----------
+        engine : {"patsy", "formulaic"} or None
+            The formula engine to use. If None, the default engine, which appears
+            in the attribute statsmodels.formula.formula_options.engine, is used.
+
+        Returns
+        -------
+        engine : {"patsy", "formulaic"}
+            The selected engine
+
+
+        Raises
+        ------
+        ValueError
+            If the selected engine is not available.
+        """
+        # Patsy for now, to be changed to a user-settable variable before release
+        _engine: Literal["patsy", "formulaic"]
+
+        if engine is not None:
+            _engine = engine
+        else:
+            import statsmodels.formula
+
+            _engine = statsmodels.formula.options.formula_engine
+
+        assert _engine is not None
+        if _engine not in ("patsy", "formulaic"):
+            raise ValueError(
+                f"Unknown engine: {_engine}. Only patsy and formulaic are supported."
+            )
+        # Ensure selected engine is available
+        msg_base = " is not available. Please install patsy."
+        if _engine == "patsy" and not HAVE_PATSY:
+            raise ImportError(f"patsy {msg_base}.")
+        if _engine == "formulaic" and not HAVE_FORMULAIC:
+            raise ImportError(f"formulaic {msg_base}.")
+
+        return _engine
+
+    @property
+    def engine(self):
+        """Get the formula engine."""
+        return self._engine
+
+    @property
+    def spec(self):
+        """
+        Get the model specification. Only available after calling get_arrays.
+        """
+        return self._spec
+
+    @property
+    def factor_evaluation_error(self):
+        if self._engine == "patsy":
+            return patsy.PatsyError
+        else:
+            return formulaic.errors.FactorEvaluationError
+
+    @property
+    def formula_materializer_error(self):
+        if self._engine == "patsy":
+            return patsy.PatsyError
+        else:
+            return formulaic.errors.FormulaMaterializerNotFoundError
+
+    @property
+    def missing_mask(self):
+        return self._missing_mask
+
+    def _legacy_orderer(
```

Comment:
1. If this is trying to reproduce `patsy`'s ordering, you should be able to just do: `Formula(<formula>).get_model_matrix(data, cluster_by="numerical_factors")`

Reply:
So far this exactly reproduced patsy's ordering in every specification I have tried.  I'll run it through the test suite and see if I can match.